tests/ckeditor.jsx

```diff
@@ -464,67 +444,138 @@ describe( 'CKEditor Component', () => {
+					res();
+
+					return Promise.resolve();
 				} );
+
+				wrapper.unmount();
 			} );
+
+			wrapper = null;
+
+			expect( editorInstance.destroy.calledOnce ).to.be.true;
 		} );
-	} );
 
-	describe( 'destroy', () => {
-		it( 'calls "Editor#destroy()" method during unmounting the component', done => {
+		it( 'should set to "null" the "editor" property inside the component', async () => {
 			const editorInstance = new Editor();
 
-			sandbox.stub( Editor, 'create' ).resolves( editorInstance );
-			sandbox.stub( editorInstance, 'destroy' ).resolves();
+			sinon.stub( Editor, 'create' ).resolves( editorInstance );
+			sinon.stub( editorInstance, 'destroy' ).resolves();
 
-			wrapper = mount( <CKEditor editor={ Editor } /> );
+			await new Promise( ( res, rej ) => {
+				wrapper = mount( <CKEditor
+					editor={ Editor }
+					onInit={ res }
+					onError={ rej } /> );
+			} );
 
-			setTimeout( () => {
-				wrapper.unmount();
-				wrapper = null;
+			const component = wrapper.instance();
 
-				expect( editorInstance.destroy.calledOnce ).to.be.true;
+			expect( component.editor ).is.not.null;
 
-				done();
+			wrapper.unmount();
+			wrapper = null;
+
+			// Wait a cycle.
+			await new Promise( res => setTimeout( res ) );
+
+			expect( component.editor ).is.null;
+		} );
+	} );
+
+	describe( 'watchdog', () => {
+		it( 'should be available for the component', async () => {
+			await new Promise( ( res, rej ) => {
+				wrapper = mount( <CKEditor
+					editor={ Editor }
+					onInit={ res }
+					onError={ rej } /> );
 			} );
+
+			expect( wrapper.instance().watchdog ).to.be.an( 'object' );
 		} );
 
-		it( 'should set to "null" the "editor" property inside the component', done => {
-			const editorInstance = new Editor();
+		it( 'should handle errors and restart the editor', async () => {
+			let onInitCallback;
+			let onErrorCallback;
 
-			sandbox.stub( Editor, 'create' ).resolves( editorInstance );
-			sandbox.stub( editorInstance, 'destroy' ).resolves();
+			function onInit() {
+				onInitCallback();
+			}
 
-			wrapper = mount( <CKEditor editor={ Editor } /> );
+			function onError() {
+				onErrorCallback();
+			}
 
-			setTimeout( () => {
-				const component = wrapper.instance();
+			const onInitSpy = sinon.spy( onInit );
+			const onErrorSpy = sinon.spy( onError );
 
-				expect( component.editor ).is.not.null;
+			wrapper = mount( <CKEditor
+				editor={ Editor }
+				onInit={ onInitSpy }
+				onError={ onErrorSpy } /> );
 
-				wrapper.unmount();
-				wrapper = null;
+			await new Promise( ( res, rej ) => {
+				onInitCallback = res;
+				onErrorCallback = rej;
+			} );
 
-				setTimeout( () => {
-					expect( component.editor ).is.null;
+			const watchdog = wrapper.instance().watchdog;
 
-					done();
-				} );
+			await new Promise( ( res, rej ) => {
+				onInitCallback = res;
+				onErrorCallback = rej;
+
+				watchdog._restart();
 			} );
+
+			sinon.assert.calledTwice( onInitSpy );
```

Comment:
I believe `onInit` should be called once. From the user point of view, the Watchdog restart mechanism is transparent. The user does not know that the editor failed and it was restarted.

I would be surprised that the `onInit()` hook was called more than once.

Reply:
Maybe the method name isn't accurate enough. In other components, there's a `ready` event.

But the second call of such event/callback is necessary in order to attach some listeners to the newly created editor, etc.

Reply:
You're right but from the end-user point of view, it does not make any sense to call the handler more than once. We need to figure out how to disable the `onInit()` callback after the first calling.

If we can't use `once()`, we need to use an additional flag.

Reply:
> You're right but from the end-user point of view, it does not make any sense to call the handler more than once.

I disagree. It depends...

Reply:
> I disagree. It depends...

From what?

Reply:
It depends on the usage. Tell me what you'd do on the `onInit` call. Anyway, that won't be a regression as earlier the editor would just crash.

Reply:
Maybe I could introduce the `onReady` event for both - editor and context? If the semantics isn't clear with the `onInit`.

Reply:
> Tell me what you'd do on the onInit call.

We can't assume that it makes sense for somebody. At this moment the handler is called once. After your changes, the handler can be called more than once which is a major breaking change that isn't mentioned in the commit message.

A first example, what I could do in the handler. Let's assume my page has a loading spinner that covers the editor area. I would like to remove the spinner when the editor is ready. After calling the handler for the second time, the script will try to remove the non-existing element and most probably end with an error.

Reply:
> Anyway, that won't be a regression as earlier the editor would just crash.

The user knew about the crash. He knew he must refresh the page and was aware that when refreshing, all callbacks will be called once again.

Watchdog "refresh" the editor in the background and calls functions that aren't supposed to be called.

Reply:
However, there is also a counter use case: what about Decoupled Editor and its UI? It must be added each time when the editor is ready.

Reply:
We agreed to change `onInit` to `onReady` and call it multiple times.